src/core/common/time_ticker.hpp

```diff
@@ -0,0 +1,136 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for a time ticker.
+ */
+
+#ifndef TIME_TICKER_HPP_
+#define TIME_TICKER_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "common/time.hpp"
+#include "common/timer.hpp"
+
+namespace ot {
+
+/**
+ * This class represents a time ticker.
+ *
+ * The time ticker emits periodic ticks (with 1 second period interval) to a set of registered tick receiver modules.
+ * The tick receivers (OpenThread objects) are identified by the `Receiver` enumeration. The receiver objects
+ * must provide `HandleTimeTick()` method which would be invoked by `TimeTicker` periodically.
+ *
+ */
+class TimeTicker : public InstanceLocator, private NonCopyable
```

Comment:
I prefer not to use `Timer` term for this since it's not really a `Timer` (the most basic operation of a timer is to allow caller to schedule it). It's an event emitter that we a user register with and then receives tick events. I was considering `TickEmitter` but decided on `TimeTicker`.

Also even though it's now limited to one second tick, we can extend it in future (allow receivers to request different tick intervals (e.g., 60 seconds ticks, etc)).  We already have a use for a 60 second tick in `network_data_leader_ftd.hpp`

Reply:
I think it's fine to keep the current name and maybe add per minute ticker in the future. 

Currently TimeTicker defines the interval to 1 second in the Doxygen. 
I wonder if there is a more explicit way to define it. 

Timer related code in DUA or MLR do requires the tick interval to be 1 second to be correct. 
```
if ((mDelay.mFields.mRegistrationDelay > 0) && (--mDelay.mFields.mRegistrationDelay == 0)) {
  attampt = true;
}
```
Otherwise this code is just correct by chance (the tick period happened to be 1s).